File: LectureCodes/NumQuad/traperr/Matlab/traperr.m
function traperr()

clear a;
global a;
l = 0; r = 0.5; % integration interval
N = 50;
a = 0.5; res05 = trapezoidal(@issin,l,r,1:N);
ex05 = trapezoidal(@issin,l,r,500); ex05 = ex05(1,2);
a = 0.9; res09 = trapezoidal(@issin,l,r,1:N);
ex09 = trapezoidal(@issin,l,r,500); ex09 = ex09(1,2);
a = 0.95; res95 = trapezoidal(@issin,l,r,1:N);
ex95 = trapezoidal(@issin,l,r,500); ex95 = ex95(1,2);
a = 0.99; res99 = trapezoidal(@issin,l,r,1:N);
ex99 = trapezoidal(@issin,l,r,500); ex99 = ex99(1,2);
figure('name','trapezoidal rule for non-periodic function');    
loglog(1./res05(:,1),abs(res05(:,2)-ex05),'r+-',...
     1./res09(:,1),abs(res09(:,2)-ex09),'b+-',...
     1./res95(:,1),abs(res95(:,2)-ex95),'c+-',...
     1./res99(:,1),abs(res99(:,2)-ex99),'m+-');
set(gca,'fontsize',12);
legend('a=0.5','a=0.9','a=0.95','a=0.99',3);
xlabel('{\bf no. of. quadrature nodes}','fontsize',14);
ylabel('{\bf |quadrature error|}','fontsize',14);
title('{\bf Trapezoidal rule quadrature for non-periodic function}','fontsize',12);

print -depsc2 '../PICTURES/traperr2.eps';

clear a;
global a;
l = 0; r = 1; % integration interval
N = 20;
a = 0.5; res05 = trapezoidal(@issin,l,r,1:N);
ex05 = trapezoidal(@issin,l,r,500); ex05 = ex05(1,2);
a = 0.9; res09 = trapezoidal(@issin,l,r,1:N);
ex09 = trapezoidal(@issin,l,r,500); ex09 = ex09(1,2);
a = 0.95; res95 = trapezoidal(@issin,l,r,1:N);
ex95 = trapezoidal(@issin,l,r,500); ex95 = ex95(1,2);
a = 0.99; res99 = trapezoidal(@issin,l,r,1:N);
ex99 = trapezoidal(@issin,l,r,500); ex99 = ex99(1,2);
figure('name','trapezoidal rule for periodic function');    
semilogy(1./res05(:,1),abs(res05(:,2)-ex05),'r+-',...
	 1./res09(:,1),abs(res09(:,2)-ex09),'b+-',...
	 1./res95(:,1),abs(res95(:,2)-ex95),'c+-',...
	 1./res99(:,1),abs(res99(:,2)-ex99),'m+-');
set(gca,'fontsize',12);
legend('a=0.5','a=0.9','a=0.95','a=0.99',3);
xlabel('{\bf no. of. quadrature nodes}','fontsize',14);
ylabel('{\bf |quadrature error|}','fontsize',14);
title('{\bf Trapezoidal rule quadrature for 1./sqrt(1-a*sin(2*pi*x+1))}','fontsize',12);

print -depsc2 '../PICTURES/traperr1.eps';

	 

File: LectureCodes/NumQuad/traperr/Eigen/traperr.cpp
# include <string>
# include <cmath>
# include <Eigen/Dense>
# include <figure/figure.hpp>
# include "trapezoidal.hpp"

void cvg(const double l, const double r, const bool logx, const std::string title, const std::string saveas) {

  // repeating the experiment for different parameters 'a'
  Eigen::VectorXd a(4); a << 0.5, 0.9, 0.95, 0.99;
  // applying trapezoidal rule for 1..N intervals
  const unsigned N = 30;

  mgl::Figure fig;
  fig.title(title);
  fig.xlabel("no. of quadrature nodes");
  fig.ylabel("|quadrature error|");
  if (logx) {
    fig.setlog(true, true); // loglog scale
  } else {
    fig.setlog(false, true); // semilogy scale
  }
  std::vector<std::string> plotstyles = {"b+", "r+", "g+", "c+"};
  for (unsigned i = 0; i < a.size(); ++i) {
    // we must define the function here, because the trapezoidal function requires a function
    // that takes only one argument! so we can't give the parameter 'a' as argument and 
    // therefore must define the function again for every different 'a' we use
    auto issin = [a, i](double x) {
      // f(x) = 1/sqrt(1 - a*sin(2*pi*x + 1)
      return 1/std::sqrt(1 - a(i)*std::sin(2*M_PI*x + 1));
    };

    std::vector<double> res; // saving results here
    res.reserve(N);
    // compute approximation for exact value by choosing a large no. of steps
    const double ex = trapezoidal(issin, l, r, N*200);

    for (unsigned n = 1; n <= N; ++n) {
      // computing error for 1..N steps
      res.push_back( std::abs(ex - trapezoidal(issin, l, r, n)) );
    }
    // plot error
    fig.plot(res, plotstyles[i]).label("a = " + std::to_string(a(i)));
  }

  fig.legend(0, 0); // activate legend and put it in bottom left corner
  fig.save(saveas);  // save plot

  return;
}

int main () {
  // test for non periodic (using loglog scale)
  cvg(0, 0.5, true, "Non periodic: on [0, 0.5]", "traperr2.eps");
  // test for periodic interval (using semilogy scale)
  cvg(0, 1, false, "Periodic: on [0,1]", "traperr1.eps");

  return 0;
}





